### PR TITLE
On 64 bit platforms, "stelem.ref" and "ldelema" ignore the high bits of a native int index

### LINKED ISSUE #52817: On x64 platform, "stelem.ref" ignores high bits of "native int" index
`IndexOutOfRangeException` not thrown when index=0x100000000.
[sharplab test](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoIgBgAIiyA6AJXwwEsBbGZgYQjcADpwA2MKAGVJAN05gYAZwDc1akQDMjMkkYl6/agG9q9c/QDaAWRgYAFhAAmASRFiAFLYfO3wsQDywlwQeErMAHIQLnhinHjxAOYAlAC6ZhZMekQo9FIYMBLcAPqsMABmHhDAAFYwYBiWqfQAhlBQaPQJeBj0nJ0BtfW9si1iODDJGRYWALwAfK3tlpzNs/Sj4zBqVDPTVt6Oru5edkd+gcGcoeFRMXHdKem7mbqMufmFMCUuKB7xjWabQ6XQBfU6YM2EymLxm5gWSygKzWGzGEx2e1hOmyuWsLXiHhhM1MWJmoygiPo6zwMAA7vRBnUGpYyLRUhi4eZyYiDNS6X0eiy2RzOd1evEnDAEFT6B4xckyAAeRWaEgiuH7GafIqlCoeYEQvCShCdPA4MRiZLqrUFHW/fXtEiG42dWhWzWvACcDqRbPdWIAvtQg1QgA=)

### PR DESCRIPTION
**Pull request by davidwrighton**
Fixes #52817

### CODE REVIEW

File: src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs [hunk 1]
```diff
@@ -596,6 +596,56 @@ private static void StelemRef(Array array, int index, object? obj)
...
+                goto assigningNull;
+
+            if (elementType != RuntimeHelpers.GetMethodTable(obj))
+                goto notExactMatch;
+
+            doWrite:
+                WriteBarrier(ref element, obj);
+                return;
+
+            assigningNull:
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs:636]: There is only one branch to this, does it need to be a labeled jump? In fact, I don't see the reason for any of these labels. Is there a benefit?

File: src/tests/JIT/Directed/Arrays/nintindexoutofrange.csproj [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <CLRTestPriority>0</CLRTestPriority>
+  </PropertyGroup>
+  <PropertyGroup>
+    <DebugType>PdbOnly</DebugType>
+  </PropertyGroup>
```
Comment (AaronRobinsonMSFT) [inline src/tests/JIT/Directed/Arrays/nintindexoutofrange.csproj:8]:
```suggestion
```

File: src/tests/JIT/Directed/Arrays/nintindexoutofrange.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    public static int Main()
+    {
+        long longIndex = ((long)1) << 32;
+        nint index = (nint)longIndex;
+        bool failed = false;
+
+        // On a 32bit platform, just succeed.
+        if ((long)index != longIndex)
```
Comment (AaronRobinsonMSFT) [inline src/tests/JIT/Directed/Arrays/nintindexoutofrange.cs:30]:
```suggestion
        if (sizeof(long) != sizeof(nint))
```

Nit: The above seems clearer to me.

**Review by AaronRobinsonMSFT — COMMENTED** [review]
Do we need to update Mono too?

**Comment by jkotas** [comment]
We had the same problem for NewArray helper some time ago and we fixed it there by changing the NewArray helper to take native int. Would it be better to do the same thing here to avoid the code duplication?

For R2R, we have introduced a new compat band in .NET 7, so it is ok to make JIT helper breaking changes without doing anything special about it.

**Review by jkotas — APPROVED** [review]
LGTM otherwise. Thank you!

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: src/tests/issues.targets [hunk 4]
```diff
@@ -1454,6 +1454,9 @@
 
     <!-- Known failures for mono runtime on *all* architectures/operating systems in *all* runtime modes -->
     <ItemGroup Condition="'$(RuntimeFlavor)' == 'mono'" >
+        <ExcludeList Include = "$(XUnitTestBinBase)/JIT/Directed/Arrays/nintindexoutofrange/**">
+            <Issue>https://github.com/dotnet/runtime/issues/71571</Issue>
```
Comment (MichalStrehovsky) [inline src/tests/issues.targets:1458]:
```suggestion
            <Issue>https://github.com/dotnet/runtime/issues/71656</Issue>
```

**Pull request merged**